// @pipecat-ai/client-react/dist/index.module.js
import {jsx as $h9lXz$jsx, Fragment as $h9lXz$Fragment} from "react/jsx-runtime";
import {RTVIEvent as $h9lXz$RTVIEvent, setAboutClient as $h9lXz$setAboutClient} from "@pipecat-ai/client-js";
import $h9lXz$react, {useRef as $h9lXz$useRef, useEffect as $h9lXz$useEffect, useCallback as $h9lXz$useCallback, useContext as $h9lXz$useContext, createContext as $h9lXz$createContext, useState as $h9lXz$useState, forwardRef as $h9lXz$forwardRef} from "react";
import {atom as $h9lXz$atom, useAtomValue as $h9lXz$useAtomValue, createStore as $h9lXz$createStore} from "jotai";
import {atomFamily as $h9lXz$atomFamily, useAtomCallback as $h9lXz$useAtomCallback} from "jotai/utils";
import {Provider as $h9lXz$Provider} from "jotai/react";

/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 


/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 



/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 





var $ad20387e24e513d4$exports = {};
$ad20387e24e513d4$exports = JSON.parse("{\"name\":\"@pipecat-ai/client-react\",\"version\":\"1.1.0\",\"license\":\"BSD-2-Clause\",\"main\":\"dist/index.js\",\"module\":\"dist/index.module.js\",\"types\":\"dist/index.d.ts\",\"source\":\"src/index.ts\",\"repository\":{\"type\":\"git\",\"url\":\"git+https://github.com/pipecat-ai/pipecat-client-web.git\"},\"files\":[\"dist\",\"package.json\",\"README.md\"],\"scripts\":{\"build\":\"parcel build --no-cache\",\"dev\":\"parcel watch\",\"lint\":\"eslint . --report-unused-disable-directives --max-warnings 0 --ignore-pattern 'dist/'\"},\"devDependencies\":{\"@pipecat-ai/client-js\":\"*\",\"@types/react\":\"^18.3.3\",\"@types/react-dom\":\"^18.3.0\",\"@typescript-eslint/eslint-plugin\":\"^8.32.0\",\"eslint\":\"^9.11.1\",\"eslint-config-prettier\":\"^9.1.0\",\"eslint-plugin-react-hooks\":\"^5.2.0\",\"eslint-plugin-simple-import-sort\":\"^12.1.1\",\"parcel\":\"^2.12.0\",\"react\":\"^18.3.1\",\"react-dom\":\"^18.3.1\",\"typescript\":\"^5.2.2\"},\"peerDependencies\":{\"@pipecat-ai/client-js\":\"*\",\"react\":\">=18\",\"react-dom\":\">=18\"},\"dependencies\":{\"jotai\":\"^2.9.0\"}}");






/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $33f16f27e37c90bc$export$fe23d0ef95286467 = /*#__PURE__*/ (0, $h9lXz$createContext)({
    on: ()=>{},
    off: ()=>{}
});


const $824ea64b5f757259$export$33a6ac53b8f02625 = (event, handler)=>{
    const { on: on, off: off } = (0, $h9lXz$useContext)((0, $33f16f27e37c90bc$export$fe23d0ef95286467));
    (0, $h9lXz$useEffect)(()=>{
        on(event, handler);
        return ()=>{
            off(event, handler);
        };
    }, [
        event,
        handler,
        on,
        off
    ]);
};


const $a90aa7250c094218$export$d6bdcccacef16204 = /*#__PURE__*/ (0, $h9lXz$createContext)({
    enableCam: ()=>{
        throw new Error("PipecatClientCamStateContext: enableCam() called outside of provider");
    },
    isCamEnabled: false
});
const $a90aa7250c094218$export$802b42df0e0d8153 = /*#__PURE__*/ (0, $h9lXz$createContext)({
    enableMic: ()=>{
        throw new Error("PipecatClientMicStateContext: enableMic() called outside of provider");
    },
    isMicEnabled: false
});
const $a90aa7250c094218$export$8e633e67c760098b = /*#__PURE__*/ (0, $h9lXz$createContext)({
    enableScreenShare: ()=>{
        throw new Error("PipecatClientScreenShareStateContext: enableScreenShare() called outside of provider");
    },
    isScreenShareEnabled: false
});
const $a90aa7250c094218$export$db79fdf85ddd6b65 = /*#__PURE__*/ (0, $h9lXz$createContext)("disconnected");
const $a90aa7250c094218$export$4777554fda61c378 = ({ children: children })=>{
    const client = (0, $034a56e7ee1b7bed$export$777fa8498be78705)();
    const [isCamEnabled, setIsCamEnabled] = (0, $h9lXz$useState)(false);
    const [isMicEnabled, setIsMicEnabled] = (0, $h9lXz$useState)(false);
    const [isScreenShareEnabled, setIsScreenShareEnabled] = (0, $h9lXz$useState)(false);
    const [transportState, setTransportState] = (0, $h9lXz$useState)("disconnected");
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).TransportStateChanged, (state)=>{
        setTransportState(state);
        if (state === "initialized" && client) {
            setIsCamEnabled(client.isCamEnabled ?? false);
            setIsMicEnabled(client.isMicEnabled ?? false);
            setIsScreenShareEnabled(client.isSharingScreen ?? false);
        }
    });
    const enableCam = (0, $h9lXz$useCallback)((enabled)=>{
        setIsCamEnabled(enabled);
        client?.enableCam?.(enabled);
    }, [
        client
    ]);
    const enableMic = (0, $h9lXz$useCallback)((enabled)=>{
        setIsMicEnabled(enabled);
        client?.enableMic?.(enabled);
    }, [
        client
    ]);
    const enableScreenShare = (0, $h9lXz$useCallback)((enabled)=>{
        client?.enableScreenShare?.(enabled);
    }, [
        client
    ]);
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).ScreenTrackStarted, (_track, participant)=>{
        if (participant?.local) setIsScreenShareEnabled(true);
    });
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).ScreenTrackStopped, (_track, participant)=>{
        if (participant?.local) setIsScreenShareEnabled(false);
    });
    return (0, $h9lXz$jsx)($a90aa7250c094218$export$db79fdf85ddd6b65.Provider, {
        value: transportState,
        children: (0, $h9lXz$jsx)($a90aa7250c094218$export$d6bdcccacef16204.Provider, {
            value: {
                enableCam: enableCam,
                isCamEnabled: isCamEnabled
            },
            children: (0, $h9lXz$jsx)($a90aa7250c094218$export$802b42df0e0d8153.Provider, {
                value: {
                    enableMic: enableMic,
                    isMicEnabled: isMicEnabled
                },
                children: (0, $h9lXz$jsx)($a90aa7250c094218$export$8e633e67c760098b.Provider, {
                    value: {
                        enableScreenShare: enableScreenShare,
                        isScreenShareEnabled: isScreenShareEnabled
                    },
                    children: children
                })
            })
        })
    });
};



const $d2e362c5a07ee3c5$var$defaultStore = (0, $h9lXz$createStore)();
const $d2e362c5a07ee3c5$export$67f6d73bc6cd7bb1 = /*#__PURE__*/ (0, $h9lXz$createContext)({});
const $d2e362c5a07ee3c5$export$bb43666ced7a20d0 = ({ children: children, client: client, jotaiStore: jotaiStore = $d2e362c5a07ee3c5$var$defaultStore })=>{
    (0, $h9lXz$useEffect)(()=>{
        (0, $h9lXz$setAboutClient)({
            library: (0, $ad20387e24e513d4$exports.name),
            library_version: (0, $ad20387e24e513d4$exports.version)
        });
    }, []);
    const eventHandlersMap = (0, $h9lXz$useRef)({});
    (0, $h9lXz$useEffect)(()=>{
        if (!client) return;
        const allEvents = Object.values((0, $h9lXz$RTVIEvent)).filter((value)=>isNaN(Number(value)));
        const allHandlers = {};
        allEvents.forEach((event)=>{
            const handler = (...payload)=>{
                const handlers = eventHandlersMap.current[event];
                if (!handlers) return;
                handlers.forEach((h)=>{
                    h(...payload);
                });
            };
            allHandlers[event] = handler;
            client.on(event, handler);
        });
        return ()=>{
            allEvents.forEach((event)=>{
                client.off(event, allHandlers[event]);
            });
        };
    }, [
        client
    ]);
    const on = (0, $h9lXz$useCallback)((event, handler)=>{
        if (!eventHandlersMap.current[event]) eventHandlersMap.current[event] = new Set();
        eventHandlersMap.current[event].add(handler);
    }, []);
    const off = (0, $h9lXz$useCallback)((event, handler)=>{
        eventHandlersMap.current[event]?.delete(handler);
    }, []);
    return (0, $h9lXz$jsx)((0, $h9lXz$Provider), {
        store: jotaiStore,
        children: (0, $h9lXz$jsx)($d2e362c5a07ee3c5$export$67f6d73bc6cd7bb1.Provider, {
            value: {
                client: client
            },
            children: (0, $h9lXz$jsx)((0, $33f16f27e37c90bc$export$fe23d0ef95286467).Provider, {
                value: {
                    on: on,
                    off: off
                },
                children: (0, $h9lXz$jsx)((0, $a90aa7250c094218$export$4777554fda61c378), {
                    children: children
                })
            })
        })
    });
};
$d2e362c5a07ee3c5$export$bb43666ced7a20d0.displayName = "PipecatClientProvider";


const $034a56e7ee1b7bed$export$777fa8498be78705 = ()=>{
    const { client: client } = (0, $h9lXz$useContext)((0, $d2e362c5a07ee3c5$export$67f6d73bc6cd7bb1));
    return client;
};



const $4b4b9099cdb5b776$var$localAudioTrackAtom = (0, $h9lXz$atom)(null);
const $4b4b9099cdb5b776$var$localVideoTrackAtom = (0, $h9lXz$atom)(null);
const $4b4b9099cdb5b776$var$localScreenAudioTrackAtom = (0, $h9lXz$atom)(null);
const $4b4b9099cdb5b776$var$localScreenVideoTrackAtom = (0, $h9lXz$atom)(null);
const $4b4b9099cdb5b776$var$botAudioTrackAtom = (0, $h9lXz$atom)(null);
const $4b4b9099cdb5b776$var$botVideoTrackAtom = (0, $h9lXz$atom)(null);
const $4b4b9099cdb5b776$var$trackAtom = (0, $h9lXz$atomFamily)(({ local: local, trackType: trackType })=>{
    if (local) switch(trackType){
        case "audio":
            return $4b4b9099cdb5b776$var$localAudioTrackAtom;
        case "screenAudio":
            return $4b4b9099cdb5b776$var$localScreenAudioTrackAtom;
        case "screenVideo":
            return $4b4b9099cdb5b776$var$localScreenVideoTrackAtom;
        case "video":
            return $4b4b9099cdb5b776$var$localVideoTrackAtom;
    }
    return trackType === "audio" ? $4b4b9099cdb5b776$var$botAudioTrackAtom : $4b4b9099cdb5b776$var$botVideoTrackAtom;
});
const $4b4b9099cdb5b776$export$9813dcd2d0c26814 = (trackType, participantType)=>{
    const client = (0, $034a56e7ee1b7bed$export$777fa8498be78705)();
    const track = (0, $h9lXz$useAtomValue)($4b4b9099cdb5b776$var$trackAtom({
        local: participantType === "local",
        trackType: trackType
    }));
    const updateTrack = (0, $h9lXz$useAtomCallback)((0, $h9lXz$useCallback)((get, set, track, trackType, local)=>{
        const atom = $4b4b9099cdb5b776$var$trackAtom({
            local: local,
            trackType: trackType
        });
        const oldTrack = get(atom);
        if (oldTrack?.id === track.id) return;
        set(atom, track);
    }, []));
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).TrackStarted, (0, $h9lXz$useCallback)((track, participant)=>{
        updateTrack(track, track.kind, Boolean(participant?.local));
    }, [
        updateTrack
    ]));
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).ScreenTrackStarted, (0, $h9lXz$useCallback)((track, participant)=>{
        const trackType = track.kind === "audio" ? "screenAudio" : "screenVideo";
        updateTrack(track, trackType, Boolean(participant?.local));
    }, [
        updateTrack
    ]));
    (0, $h9lXz$useEffect)(()=>{
        if (!client) return;
        const tracks = client.tracks();
        const track = tracks?.[participantType]?.[trackType];
        if (!track) return;
        updateTrack(track, trackType, participantType === "local");
    }, [
        participantType,
        trackType,
        updateTrack,
        client
    ]);
    return track;
};



const $f209aa7ddb77dcb2$export$b52250cb73ff4de1 = ()=>{
    const botAudioRef = (0, $h9lXz$useRef)(null);
    const botAudioTrack = (0, $4b4b9099cdb5b776$export$9813dcd2d0c26814)("audio", "bot");
    (0, $h9lXz$useEffect)(()=>{
        if (!botAudioRef.current || !botAudioTrack) return;
        if (botAudioRef.current.srcObject) {
            const oldTrack = botAudioRef.current.srcObject.getAudioTracks()[0];
            if (oldTrack.id === botAudioTrack.id) return;
        }
        botAudioRef.current.srcObject = new MediaStream([
            botAudioTrack
        ]);
    }, [
        botAudioTrack
    ]);
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).SpeakerUpdated, (0, $h9lXz$useCallback)((speaker)=>{
        if (!botAudioRef.current) return;
        if (typeof botAudioRef.current.setSinkId !== "function") return;
        botAudioRef.current.setSinkId(speaker.deviceId);
    }, []));
    return (0, $h9lXz$jsx)((0, $h9lXz$Fragment), {
        children: (0, $h9lXz$jsx)("audio", {
            ref: botAudioRef,
            autoPlay: true
        })
    });
};
$f209aa7ddb77dcb2$export$b52250cb73ff4de1.displayName = "PipecatClientAudio";




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $e76ee2f021b54325$export$3ea2601427f0430f = ()=>(0, $h9lXz$useContext)((0, $a90aa7250c094218$export$d6bdcccacef16204));


const $7cb2ce2c4cbfb401$export$dc9a029eeca8213f = ({ onCamEnabledChanged: onCamEnabledChanged, disabled: disabled = false, children: children })=>{
    const { isCamEnabled: isCamEnabled, enableCam: enableCam } = (0, $e76ee2f021b54325$export$3ea2601427f0430f)();
    const handleToggleCam = (0, $h9lXz$useCallback)(()=>{
        if (disabled) return;
        const newEnabledState = !isCamEnabled;
        enableCam(newEnabledState);
        onCamEnabledChanged?.(newEnabledState);
    }, [
        disabled,
        enableCam,
        isCamEnabled,
        onCamEnabledChanged
    ]);
    return (0, $h9lXz$jsx)((0, $h9lXz$Fragment), {
        children: children({
            isCamEnabled: isCamEnabled,
            onClick: handleToggleCam,
            disabled: disabled
        })
    });
};
var $7cb2ce2c4cbfb401$export$2e2bcd8739ae039 = $7cb2ce2c4cbfb401$export$dc9a029eeca8213f;




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $5905c001b0dc8d25$export$388e706586309ef0 = ()=>(0, $h9lXz$useContext)((0, $a90aa7250c094218$export$802b42df0e0d8153));


const $2984fdfc31bad375$export$bc8133b69ff660a2 = ({ onMicEnabledChanged: onMicEnabledChanged, disabled: disabled = false, children: children })=>{
    const { enableMic: enableMic, isMicEnabled: isMicEnabled } = (0, $5905c001b0dc8d25$export$388e706586309ef0)();
    const handleToggleMic = (0, $h9lXz$useCallback)(()=>{
        if (disabled) return;
        const newEnabledState = !isMicEnabled;
        enableMic(newEnabledState);
        onMicEnabledChanged?.(newEnabledState);
    }, [
        disabled,
        enableMic,
        isMicEnabled,
        onMicEnabledChanged
    ]);
    return (0, $h9lXz$jsx)((0, $h9lXz$Fragment), {
        children: children({
            isMicEnabled: isMicEnabled,
            onClick: handleToggleMic,
            disabled: disabled
        })
    });
};
var $2984fdfc31bad375$export$2e2bcd8739ae039 = $2984fdfc31bad375$export$bc8133b69ff660a2;





/**
 * Copyright (c) 2025, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $d39c0bc3744ea030$export$be63b19bd7f7d4f5 = ()=>(0, $h9lXz$useContext)((0, $a90aa7250c094218$export$8e633e67c760098b));


const $d6dc7e80e6e976a0$export$93764714dfab3a46 = ({ onScreenShareEnabledChanged: onScreenShareEnabledChanged, disabled: disabled = false, children: children })=>{
    const { enableScreenShare: enableScreenShare, isScreenShareEnabled: isScreenShareEnabled } = (0, $d39c0bc3744ea030$export$be63b19bd7f7d4f5)();
    const handleToggleScreenShare = (0, $h9lXz$useCallback)(()=>{
        if (disabled) return;
        enableScreenShare(!isScreenShareEnabled);
    }, [
        disabled,
        enableScreenShare,
        isScreenShareEnabled
    ]);
    (0, $h9lXz$useEffect)(()=>{
        onScreenShareEnabledChanged?.(isScreenShareEnabled);
    }, [
        isScreenShareEnabled,
        onScreenShareEnabledChanged
    ]);
    return (0, $h9lXz$jsx)((0, $h9lXz$Fragment), {
        children: children({
            isScreenShareEnabled: isScreenShareEnabled,
            onClick: handleToggleScreenShare,
            disabled: disabled
        })
    });
};
var $d6dc7e80e6e976a0$export$2e2bcd8739ae039 = $d6dc7e80e6e976a0$export$93764714dfab3a46;




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * This file contains code derived from:
 * https://github.com/jaredLunde/react-hook/blob/master/packages/merged-ref/src/index.tsx
 * Original author: Jared Lunde (https://github.com/jaredLunde)
 * Original license: MIT (https://github.com/jaredLunde/react-hook/blob/master/LICENSE)
 */ 
function $9098519210cf34e2$var$useMergedRef(...refs) {
    return (0, $h9lXz$useCallback)((element)=>{
        for(let i = 0; i < refs.length; i++){
            const ref = refs[i];
            if (typeof ref === "function") ref(element);
            else if (ref && typeof ref === "object") ref.current = element;
        }
    }, // eslint-disable-next-line react-hooks/exhaustive-deps
    refs);
}
var $9098519210cf34e2$export$2e2bcd8739ae039 = $9098519210cf34e2$var$useMergedRef;



const $6a65deb8615a2ad7$export$85974db6d0cc43b3 = /*#__PURE__*/ (0, $h9lXz$forwardRef)(function VoiceClientVideo({ participant: participant = "local", fit: fit = "contain", mirror: mirror, onResize: onResize, style: style = {}, trackType: trackType = "video", ...props }, ref) {
    const videoTrack = (0, $4b4b9099cdb5b776$export$9813dcd2d0c26814)(trackType, participant);
    const videoEl = (0, $h9lXz$useRef)(null);
    const videoRef = (0, $9098519210cf34e2$export$2e2bcd8739ae039)(videoEl, ref);
    /**
     * Handle canplay & picture-in-picture events.
     */ (0, $h9lXz$useEffect)(function setupVideoEvents() {
        const video = videoEl.current;
        if (!video) return;
        const playVideo = ()=>{
            const promise = video.play();
            if (promise !== undefined) promise.then(()=>{
                // All good, playback started.
                video.controls = false;
            }).catch((error)=>{
                // Auto-play was prevented. Show video controls, so user can play video manually.
                video.controls = true;
                console.warn("Failed to play video", error);
            });
        };
        const handleCanPlay = ()=>{
            if (!video.paused) return;
            playVideo();
        };
        const handleEnterPIP = ()=>{
            video.style.transform = "scale(1)";
        };
        const handleLeavePIP = ()=>{
            video.style.transform = "";
            setTimeout(()=>{
                if (video.paused) playVideo();
            }, 100);
        };
        const handleVisibilityChange = ()=>{
            if (document.visibilityState === "hidden") return;
            if (!video.paused) return;
            playVideo();
        };
        video.addEventListener("canplay", handleCanPlay);
        video.addEventListener("enterpictureinpicture", handleEnterPIP);
        video.addEventListener("leavepictureinpicture", handleLeavePIP);
        // Videos can be paused if media was played in another app on iOS.
        document.addEventListener("visibilitychange", handleVisibilityChange);
        return ()=>{
            video.removeEventListener("canplay", handleCanPlay);
            video.removeEventListener("enterpictureinpicture", handleEnterPIP);
            video.removeEventListener("leavepictureinpicture", handleLeavePIP);
            document.removeEventListener("visibilitychange", handleVisibilityChange);
        };
    }, []);
    /**
     * Update srcObject.
     */ (0, $h9lXz$useEffect)(function updateSrcObject() {
        const video = videoEl.current;
        if (!video || !videoTrack) return;
        video.srcObject = new MediaStream([
            videoTrack
        ]);
        video.load();
        return ()=>{
            // clean up when unmounted
            video.srcObject = null;
            video.load();
        };
    }, [
        videoTrack,
        videoTrack?.id
    ]);
    /**
     * Add optional event listener for resize event so the parent component
     * can know the video's native aspect ratio.
     */ (0, $h9lXz$useEffect)(function reportVideoDimensions() {
        const video = videoEl.current;
        if (!onResize || !video) return;
        let frame;
        function handleResize() {
            if (frame) cancelAnimationFrame(frame);
            frame = requestAnimationFrame(()=>{
                const video = videoEl.current;
                if (!video || document.hidden) return;
                const videoWidth = video.videoWidth;
                const videoHeight = video.videoHeight;
                if (videoWidth && videoHeight) onResize?.({
                    aspectRatio: videoWidth / videoHeight,
                    height: videoHeight,
                    width: videoWidth
                });
            });
        }
        handleResize();
        video.addEventListener("loadedmetadata", handleResize);
        video.addEventListener("resize", handleResize);
        return ()=>{
            if (frame) cancelAnimationFrame(frame);
            video.removeEventListener("loadedmetadata", handleResize);
            video.removeEventListener("resize", handleResize);
        };
    }, [
        onResize
    ]);
    return (0, $h9lXz$jsx)("video", {
        autoPlay: true,
        muted: true,
        playsInline: true,
        ref: videoRef,
        style: {
            objectFit: fit,
            transform: mirror ? "scale(-1, 1)" : "",
            ...style
        },
        ...props
    });
});
$6a65deb8615a2ad7$export$85974db6d0cc43b3.displayName = "PipecatClientVideo";




/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 





const $f934f1f8b10aaf19$var$availableMicsAtom = (0, $h9lXz$atom)([]);
const $f934f1f8b10aaf19$var$availableCamsAtom = (0, $h9lXz$atom)([]);
const $f934f1f8b10aaf19$var$availableSpeakersAtom = (0, $h9lXz$atom)([]);
const $f934f1f8b10aaf19$var$selectedMicAtom = (0, $h9lXz$atom)({});
const $f934f1f8b10aaf19$var$selectedCamAtom = (0, $h9lXz$atom)({});
const $f934f1f8b10aaf19$var$selectedSpeakerAtom = (0, $h9lXz$atom)({});
const $f934f1f8b10aaf19$export$642bc4d2d2a376f1 = ()=>{
    const client = (0, $034a56e7ee1b7bed$export$777fa8498be78705)();
    const availableCams = (0, $h9lXz$useAtomValue)($f934f1f8b10aaf19$var$availableCamsAtom);
    const availableMics = (0, $h9lXz$useAtomValue)($f934f1f8b10aaf19$var$availableMicsAtom);
    const availableSpeakers = (0, $h9lXz$useAtomValue)($f934f1f8b10aaf19$var$availableSpeakersAtom);
    const selectedCam = (0, $h9lXz$useAtomValue)($f934f1f8b10aaf19$var$selectedCamAtom);
    const selectedMic = (0, $h9lXz$useAtomValue)($f934f1f8b10aaf19$var$selectedMicAtom);
    const selectedSpeaker = (0, $h9lXz$useAtomValue)($f934f1f8b10aaf19$var$selectedSpeakerAtom);
    const initDevices = (0, $h9lXz$useAtomCallback)((0, $h9lXz$useCallback)(async (_get, set)=>{
        if (!client) return;
        const availableCams = await client.getAllCams();
        const availableMics = await client.getAllMics();
        const availableSpeakers = await client.getAllSpeakers();
        set($f934f1f8b10aaf19$var$availableCamsAtom, availableCams);
        set($f934f1f8b10aaf19$var$availableMicsAtom, availableMics);
        set($f934f1f8b10aaf19$var$availableSpeakersAtom, availableSpeakers);
        set($f934f1f8b10aaf19$var$selectedCamAtom, client.selectedCam);
        set($f934f1f8b10aaf19$var$selectedMicAtom, client.selectedMic);
        set($f934f1f8b10aaf19$var$selectedSpeakerAtom, client.selectedSpeaker);
    }, [
        client
    ]));
    (0, $h9lXz$useEffect)(()=>{
        initDevices();
    }, [
        initDevices
    ]);
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).AvailableCamsUpdated, (0, $h9lXz$useAtomCallback)((0, $h9lXz$useCallback)((_get, set, cams)=>{
        set($f934f1f8b10aaf19$var$availableCamsAtom, cams);
    }, [])));
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).AvailableMicsUpdated, (0, $h9lXz$useAtomCallback)((0, $h9lXz$useCallback)((_get, set, mics)=>{
        set($f934f1f8b10aaf19$var$availableMicsAtom, mics);
    }, [])));
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).AvailableSpeakersUpdated, (0, $h9lXz$useAtomCallback)((0, $h9lXz$useCallback)((_get, set, speakers)=>{
        set($f934f1f8b10aaf19$var$availableSpeakersAtom, speakers);
    }, [])));
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).CamUpdated, (0, $h9lXz$useAtomCallback)((0, $h9lXz$useCallback)((_get, set, cam)=>{
        set($f934f1f8b10aaf19$var$selectedCamAtom, cam);
    }, [])));
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).MicUpdated, (0, $h9lXz$useAtomCallback)((0, $h9lXz$useCallback)((_get, set, mic)=>{
        set($f934f1f8b10aaf19$var$selectedMicAtom, mic);
    }, [])));
    (0, $824ea64b5f757259$export$33a6ac53b8f02625)((0, $h9lXz$RTVIEvent).SpeakerUpdated, (0, $h9lXz$useAtomCallback)((0, $h9lXz$useCallback)((_get, set, speaker)=>{
        set($f934f1f8b10aaf19$var$selectedSpeakerAtom, speaker);
    }, [])));
    const updateCam = (0, $h9lXz$useCallback)((id)=>{
        client?.updateCam(id);
    }, [
        client
    ]);
    const updateMic = (0, $h9lXz$useCallback)((id)=>{
        client?.updateMic(id);
    }, [
        client
    ]);
    const updateSpeaker = (0, $h9lXz$useCallback)((id)=>{
        client?.updateSpeaker(id);
    }, [
        client
    ]);
    return {
        availableCams: availableCams,
        availableMics: availableMics,
        availableSpeakers: availableSpeakers,
        selectedCam: selectedCam,
        selectedMic: selectedMic,
        selectedSpeaker: selectedSpeaker,
        updateCam: updateCam,
        updateMic: updateMic,
        updateSpeaker: updateSpeaker
    };
};





/**
 * Copyright (c) 2024, Daily.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */ 

const $33f3729bbe9f09df$export$30aee278309a867b = ()=>(0, $h9lXz$useContext)((0, $a90aa7250c094218$export$db79fdf85ddd6b65));






const $993a744193844a95$export$59bf27bd43679db6 = /*#__PURE__*/ (0, $h9lXz$react).memo(({ backgroundColor: backgroundColor = "transparent", barColor: barColor = "black", barCount: barCount = 5, barGap: barGap = 12, barLineCap: barLineCap = "round", barMaxHeight: barMaxHeight = 120, barOrigin: barOrigin = "center", barWidth: barWidth = 30, participantType: participantType })=>{
    const canvasRef = (0, $h9lXz$useRef)(null);
    const track = (0, $4b4b9099cdb5b776$export$9813dcd2d0c26814)("audio", participantType);
    (0, $h9lXz$useEffect)(()=>{
        if (!canvasRef.current) return;
        const canvasWidth = barCount * barWidth + (barCount - 1) * barGap;
        const canvasHeight = barMaxHeight;
        const canvas = canvasRef.current;
        const scaleFactor = 2;
        // Make canvas fill the width and height of its container
        const resizeCanvas = ()=>{
            canvas.width = canvasWidth * scaleFactor;
            canvas.height = canvasHeight * scaleFactor;
            canvas.style.width = `${canvasWidth}px`;
            canvas.style.height = `${canvasHeight}px`;
            canvasCtx.lineCap = barLineCap;
            canvasCtx.scale(scaleFactor, scaleFactor);
        };
        const canvasCtx = canvas.getContext("2d");
        resizeCanvas();
        if (!track) return;
        const audioContext = new AudioContext();
        const source = audioContext.createMediaStreamSource(new MediaStream([
            track
        ]));
        const analyser = audioContext.createAnalyser();
        analyser.fftSize = 1024;
        source.connect(analyser);
        const frequencyData = new Uint8Array(analyser.frequencyBinCount);
        canvasCtx.lineCap = barLineCap;
        // Create frequency bands based on barCount
        const bands = Array.from({
            length: barCount
        }, (_, i)=>{
            // Use improved logarithmic scale for better frequency distribution
            const minFreq = barCount > 20 ? 200 : 80; // Adjust min frequency based on bar count
            const maxFreq = 10000; // Cover most important audio frequencies
            // Use Mel scale inspired approach for more perceptually uniform distribution
            // This helps with a large number of bars by placing fewer in the very low range
            // https://en.wikipedia.org/wiki/Mel_scale
            const melMin = 2595 * Math.log10(1 + minFreq / 700);
            const melMax = 2595 * Math.log10(1 + maxFreq / 700);
            const melStep = (melMax - melMin) / barCount;
            const melValue = melMin + i * melStep;
            const startFreq = 700 * (Math.pow(10, melValue / 2595) - 1);
            const endFreq = 700 * (Math.pow(10, (melValue + melStep) / 2595) - 1);
            return {
                startFreq: startFreq,
                endFreq: endFreq,
                smoothValue: 0
            };
        });
        const getFrequencyBinIndex = (frequency)=>{
            const nyquist = audioContext.sampleRate / 2;
            return Math.round(frequency / nyquist * (analyser.frequencyBinCount - 1));
        };
        function drawSpectrum() {
            analyser.getByteFrequencyData(frequencyData);
            canvasCtx.clearRect(0, 0, canvas.width / scaleFactor, canvas.height / scaleFactor);
            canvasCtx.fillStyle = backgroundColor;
            canvasCtx.fillRect(0, 0, canvas.width / scaleFactor, canvas.height / scaleFactor);
            let isActive = false;
            const totalBarsWidth = bands.length * barWidth + (bands.length - 1) * barGap;
            const startX = (canvas.width / scaleFactor - totalBarsWidth) / 2; // Center bars
            const adjustedCircleRadius = barWidth / 2; // Fixed radius for reset circles
            bands.forEach((band, i)=>{
                const startIndex = getFrequencyBinIndex(band.startFreq);
                const endIndex = getFrequencyBinIndex(band.endFreq);
                const bandData = frequencyData.slice(startIndex, endIndex);
                const bandValue = bandData.reduce((acc, val)=>acc + val, 0) / bandData.length;
                const smoothingFactor = 0.2;
                if (bandValue < 1) band.smoothValue = Math.max(band.smoothValue - smoothingFactor * 5, 0);
                else {
                    band.smoothValue = band.smoothValue + (bandValue - band.smoothValue) * smoothingFactor;
                    isActive = true;
                }
                const x = startX + i * (barWidth + barGap);
                // Calculate bar height with a maximum cap
                const minHeight = 0;
                const barHeight = Math.max(minHeight, Math.min(band.smoothValue / 255 * barMaxHeight, barMaxHeight));
                let yTop, yBottom;
                const canvasHeight = canvas.height / scaleFactor;
                switch(barOrigin){
                    case "top":
                        yTop = adjustedCircleRadius;
                        yBottom = Math.min(adjustedCircleRadius + barHeight, canvasHeight - adjustedCircleRadius);
                        break;
                    case "bottom":
                        yBottom = canvasHeight - adjustedCircleRadius;
                        yTop = Math.max(yBottom - barHeight, adjustedCircleRadius);
                        break;
                    case "center":
                    default:
                        yTop = Math.max(canvasHeight / 2 - barHeight / 2, adjustedCircleRadius);
                        yBottom = Math.min(canvasHeight / 2 + barHeight / 2, canvasHeight - adjustedCircleRadius);
                        break;
                }
                if (band.smoothValue > 0) {
                    canvasCtx.beginPath();
                    canvasCtx.moveTo(x + barWidth / 2, yTop);
                    canvasCtx.lineTo(x + barWidth / 2, yBottom);
                    canvasCtx.lineWidth = barWidth;
                    canvasCtx.strokeStyle = barColor;
                    canvasCtx.stroke();
                } else drawInactiveCircle(adjustedCircleRadius, barColor, x, yTop);
            });
            if (!isActive) drawInactiveCircles(adjustedCircleRadius, barColor);
            requestAnimationFrame(drawSpectrum);
        }
        function drawInactiveCircle(circleRadius, color, x, y) {
            switch(barLineCap){
                case "square":
                    canvasCtx.fillStyle = color;
                    canvasCtx.fillRect(x + barWidth / 2 - circleRadius, y - circleRadius, circleRadius * 2, circleRadius * 2);
                    break;
                case "round":
                default:
                    canvasCtx.beginPath();
                    canvasCtx.arc(x + barWidth / 2, y, circleRadius, 0, 2 * Math.PI);
                    canvasCtx.fillStyle = color;
                    canvasCtx.fill();
                    canvasCtx.closePath();
                    break;
            }
        }
        function drawInactiveCircles(circleRadius, color) {
            const totalBarsWidth = bands.length * barWidth + (bands.length - 1) * barGap;
            const startX = (canvas.width / scaleFactor - totalBarsWidth) / 2;
            const canvasHeight = canvas.height / scaleFactor;
            let y;
            switch(barOrigin){
                case "top":
                    y = circleRadius;
                    break;
                case "bottom":
                    y = canvasHeight - circleRadius;
                    break;
                case "center":
                default:
                    y = canvasHeight / 2;
                    break;
            }
            bands.forEach((_, i)=>{
                const x = startX + i * (barWidth + barGap);
                drawInactiveCircle(circleRadius, color, x, y);
            });
        }
        drawSpectrum();
        // Handle resizing
        window.addEventListener("resize", resizeCanvas);
        return ()=>{
            audioContext.close();
            window.removeEventListener("resize", resizeCanvas);
        };
    }, [
        backgroundColor,
        barColor,
        barCount,
        barGap,
        barLineCap,
        barMaxHeight,
        barOrigin,
        barWidth,
        track
    ]);
    return (0, $h9lXz$jsx)("canvas", {
        ref: canvasRef,
        style: {
            display: "block",
            width: "100%",
            height: "100%"
        }
    });
});
$993a744193844a95$export$59bf27bd43679db6.displayName = "VoiceVisualizer";




export {$f209aa7ddb77dcb2$export$b52250cb73ff4de1 as PipecatClientAudio, $7cb2ce2c4cbfb401$export$dc9a029eeca8213f as PipecatClientCamToggle, $2984fdfc31bad375$export$bc8133b69ff660a2 as PipecatClientMicToggle, $d2e362c5a07ee3c5$export$bb43666ced7a20d0 as PipecatClientProvider, $d6dc7e80e6e976a0$export$93764714dfab3a46 as PipecatClientScreenShareToggle, $6a65deb8615a2ad7$export$85974db6d0cc43b3 as PipecatClientVideo, $034a56e7ee1b7bed$export$777fa8498be78705 as usePipecatClient, $e76ee2f021b54325$export$3ea2601427f0430f as usePipecatClientCamControl, $f934f1f8b10aaf19$export$642bc4d2d2a376f1 as usePipecatClientMediaDevices, $4b4b9099cdb5b776$export$9813dcd2d0c26814 as usePipecatClientMediaTrack, $5905c001b0dc8d25$export$388e706586309ef0 as usePipecatClientMicControl, $d39c0bc3744ea030$export$be63b19bd7f7d4f5 as usePipecatClientScreenShareControl, $33f3729bbe9f09df$export$30aee278309a867b as usePipecatClientTransportState, $824ea64b5f757259$export$33a6ac53b8f02625 as useRTVIClientEvent, $993a744193844a95$export$59bf27bd43679db6 as VoiceVisualizer};
